sv.c

```diff
@@ -14633,6 +14633,9 @@ S_sv_dup_common(pTHX_ const SV *const ssv, CLONE_PARAMS *const param)
                 } else if (CvCONST(dsv)) {
                     CvXSUBANY(dsv).any_ptr =
                         sv_dup_inc((const SV *)CvXSUBANY(dsv).any_ptr, param);
+                } else if (CvREFCOUNTED_ANYSV(dsv)) {
+                    CvXSUBANY(dsv).any_sv =
+                        sv_dup_inc((const SV *)CvXSUBANY(dsv).any_sv, param);
```

Comment:
> Is that still the case? Should that be fixed as well? (possibly as a separate commit/PR);
> 
> i.e. should the code be updated with something like:
> 
> ```
>               else {
>                     CvXSUBANY(dsv).any_sv =
>                         sv_dup((const SV *)CvXSUBANY(dsv).any_sv, param);
>               }
> ```

Unfortunately it can't be. The `ANY` is a union, so there's no telling which field (if indeed any at all) are actually in use. In particular we can't distinguish between this being an SV pointer, some other (maybe-malloc()'ed) pointer, an integer, or just memory junk. Only the behaviour of the code itself will know.

This in fact is why I gated all the logic behind a new flag. New code sets the flag only when it is definitely using the `.any_sv` field to store an SV pointer, so we can rely on that fact. Without the flag we cannot make any assumptions.

Reply:
> There was some speculation that this code might already be broken for threads (that is if dup is called for a CV which uses CvXSUBANY).

Yeah, in `sv_dup` it will need to `sv_dup_inc`, and `cv_clone` will need to `SvREFCNT_inc`

Reply:
> Yeah, in `sv_dup` it will need to `sv_dup_inc`, and `cv_clone` will need to `SvREFCNT_inc`

Which it now does, in the newer commits.

Reply:
> Next question then: can/should/could it warn/croak/... if dup is called when `CvISXSUB(dsv)` is true?
> The goal appears to be to 'duplicate an SV of any type' but for a SV which has an `ANY` that isn't really the case since whatever is in `ANY` doesn't get duplicated..

It cannot. It's perfectly fine to `sv_dup` a CV containing an XSUB. And it cannot reliably detect if the XSUB happens to be using the CvXSUBANY union at all; and if it is using it, which field, and to what it points. There's already going to be much code which happens to do safe things, and happens to do unsafe things, and we cannot tell which is which. I don't think a warning is possible.